src/DGtal/topology/CubicalComplex.h

```diff
@@ -0,0 +1,1305 @@
+
+    template < typename TKSpace, typename TCellContainer >
+    bool
+    operator==( const CubicalComplex< TKSpace, TCellContainer >&,
+                const CubicalComplex< TKSpace, TCellContainer >& );
+    template < typename TKSpace, typename TCellContainer >
+    bool
+    operator!=( const CubicalComplex< TKSpace, TCellContainer >&,
+                const CubicalComplex< TKSpace, TCellContainer >& );
+    template < typename TKSpace, typename TCellContainer >
+    bool
+    operator<=( const CubicalComplex< TKSpace, TCellContainer >&,
+                const CubicalComplex< TKSpace, TCellContainer >& );
+    template < typename TKSpace, typename TCellContainer >
+    bool
+    operator>=( const CubicalComplex< TKSpace, TCellContainer >&,
+                const CubicalComplex< TKSpace, TCellContainer >& );
+  } // namespace functions
+
+
+  /////////////////////////////////////////////////////////////////////////////
+  // template class CubicalComplex
+  /**
+  * Description of template class 'CubicalComplex' <p> \brief Aim:
+  * This class represents an arbitrary cubical complex living in some
+  * Khalimsky space. Cubical complexes are sets of cells of different
+  * dimensions related together with incidence relations. Two cells
+  * in a cubical complex are incident if and only if they are
+  * incident in the surrounding Khalimsky space. In other words,
+  * cubical complexes are defined here as subsets of Khalimsky spaces. 
+  *
+  * A cubical complex is a (immutable) model of boost::Container and
+  * offers forward iterators to enumerate elements. It is close from
+  * being an AssociativeContainer, but values are not sorted (they are
+  * sorted per dimension), and are not modifiable. It is not exactly a
+  * container in the usual sense because it cannot be constructed by
+  * default, and because iterators may not modified values.
+  *
+  * @tparam TKSpace any model of concepts::CCellularGridSpaceND, i.e. a type
+  * that models a Khalimsky space.
+  *
+  * @tparam TCellContainer any model of associative container, mapping
+  * a KSpace::Cell to a CubicalCellData or any type deriving from
+  * it. It could be for instance a std::map or a
+  * std::unordered_map. Note that unfortunately, unordered_map are
+  * (strangely) not models of boost::AssociativeContainer, hence we
+  * cannot check concepts here.
+  *
+  */
+  template < typename TKSpace, 
+             typename TCellContainer = std::map< typename TKSpace::Cell, CubicalCellData > >
+  class CubicalComplex
+  {
+    // ----------------------- associated types ------------------------------
+  public:
+    typedef CubicalComplex< TKSpace, TCellContainer > Self;
+    
+    BOOST_CONCEPT_ASSERT(( concepts::CCellularGridSpaceND< TKSpace > ));
+    BOOST_STATIC_ASSERT( IsPairAssociativeContainer< TCellContainer >::value );
+
+    // JOL: Not used, because unordered_set and unordered_map do not pass
+    // these concept checks.
```

Comment:
I prefer not, because CSTLAssociativeContainer is not "standard" while the others are. They should be put back when they are fixed. I think the comment in the file is not a problem and explains why we check the concept like that.